client/src/redux/sagas/applicationSaga.js

```diff
@@ -23,26 +23,46 @@ function* sendApplication({ payload }) {
 function* moveApplicant({ payload }) {
   try {
     const recruiter = yield select(getCurrentUser)
+    const stages = yield select(getStages)
     const token = recruiter.token
-    const applicant = payload.applicant
-    const newStage = payload.newStage
+    const { applicant, newStage, oldStage, oldIndex } = payload
     const data = {
-      postingStageId: newStage.id,
-      jobApplicationId: applicant.id
+      postingStageId: newStage,
+      jobApplicationId: applicant
     }
 
+    const oStage = stages.find(stage => stage.id === oldStage)
+    const nStage = stages.find(stage => stage.id === newStage)
+
+    let oldApplicants = Array.from(oStage.applicants)
+    let newApplicants = Array.from(nStage.applicants)
+
+    const iterated = oldApplicants.map(a =>
+      (a.id === applicant) ?
+        { ...a, updatedAt: (new Date()).toJSON() }
+        : { ...a }
+    )
+
+    if (oldStage !== newStage) {
+      const movedApplicant = iterated.splice(oldIndex, 1)[0]
+      newApplicants.splice(0, 0, movedApplicant)
+    }
+
+    const reArrangedStages = stages.map(stage =>
+      (stage.id === oldStage) ?
+        { ...stage, applicants: iterated }
+        : (stage.id === newStage) ?
+          { ...stage, applicants: newApplicants }
+          : { ...stage }
+    )
+
+    yield put(actions.fetchApplicantsSuccess(reArrangedStages))
+
     const response = yield call(jobApplicationApi.moveApplicants, { token, data })
 
     if (response.status === 200) {
-      const stages = yield select(getStages)
-      const filteredStages = stages.map(stage => (
-        { ...stage, applicants: stage.applicants.filter(a => a.id !== applicant.id) }
-      ))
-      const finalStages = filteredStages.map(stage => (stage.id === newStage.id) ?
-        { ...stage, applicants: [...stage.applicants, applicant] }
-        : { ...stage }
-      )
-      yield put(actions.moveApplicantSuccess(finalStages))
+      const jobPosting = yield select(getCurrentJobPosting)
+      yield put(actions.fetchApplicants(jobPosting.id))
     }
```

Comment:
The saga only fetches applicants if the move was successful but always updates the store, so if the move/request fails for some reason the user is not notified and the view changes as if the move was successful.